agent/src/beerocks/slave/son_slave_thread.cpp

```diff
@@ -4440,32 +4440,41 @@ beerocks::message::sWifiChannel slave_thread::channel_selection_select_channel()
             continue;
         }
         for (uint8_t i = 0; i < beerocks::message::SUPPORTED_CHANNELS_LENGTH; i++) {
-            auto channel         = hostap_params.supported_channels[i];
-            auto operating_class = wireless_utils::get_operating_class_by_channel(
-                channel.channel, static_cast<beerocks::eWiFiBandwidth>(channel.channel_bandwidth));
+            auto channel        = hostap_params.supported_channels[i];
+            auto bw             = static_cast<beerocks::eWiFiBandwidth>(channel.channel_bandwidth);
+            auto center_channel = wireless_utils::get_5g_center_channel(channel.channel, bw, true);
+            auto operating_class =
+                wireless_utils::get_operating_class_by_channel(channel.channel, center_channel, bw);
 
             // Skip channels from other operating classes
             if (operating_class != preference.oper_class) {
                 continue;
             }
             // Skip DFS channels
-            if (channel.is_dfs_channel) {
+            if (hostap_params.supported_channels[i].is_dfs_channel) {
                 LOG(DEBUG) << "Skip DFS channel " << int(channel.channel);
                 continue;
             }
             auto is_restricted =
                 std::find_if(preference.channels.begin(), preference.channels.end(),
                              [&](const beerocks::message::sWifiChannel &wifi_channel) {
-                                 return (channel.channel == wifi_channel.channel);
+                                 // operating classes 128,129,130 use center channel **unlike the other classes** (See Table E-4 in 802.11 spec)
```

Comment:
802.11 spec say its VS ,
 so if this is our decision why not use first channel as well as the other classes?

Reply:
Don't know what you mean by "spec say its VS". When we get channel preference TLV from the controller, it includes all the non-operable channels for each operating class.
For operating classes 128,129,130 the 802.11 specification says that the center channel should be used instead of the start channel, which is also reflected in the channel set that the UCC sends in the channel selection request:
```
DEBUG 22:54:00:028 <1988852512> son_slave_thread.cpp[4566] --> operating class=128, preference=0, reason=0, channel_list={42,58,106,122,138}
DEBUG 22:54:00:028 <1988852512> son_slave_thread.cpp[4566] --> operating class=129, preference=0, reason=0, channel_list={}
DEBUG 22:54:00:029 <1988852512> son_slave_thread.cpp[4566] --> operating class=130, preference=0, reason=0, channel_list={}
```

Since operating class 128 bandwidth is 80MHz, channel 42 is the center channel while the start channel (which is what is expected/reported by hostapd) is 36.
So in this code snippet, when we compare the channel to the non-operable channels, we should compare the center channel instead of the start channel when the operating class is 128-130.

I hope this is more clear.

Reply:
didn't see it in  802.11 2012 spec

![image](https://user-images.githubusercontent.com/51229524/76430136-a31de880-63b8-11ea-9c72-a28e5b20fd05.png)

so in that case we can decide what we send in the preference TLV for operative classes 128-130